util/stop: change stopper methods to return an error

vivekmenezes opened the pull request:
So that code always considers the "stopper is stopping" error case.

fixes #7305 

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7459)

<!-- Reviewable:end -->

vivekmenezes committed:
util/stop: change stopper methods to return an error

So that code always considers the "node stopped" error case.

tschottdorf commented on gossip/server.go:
```diff
@@ -97,11 +97,13 @@ func (s *server) Gossip(stream Gossip_GossipServer) error {
 		})
-	})
+	}); err != nil {
+		return nil
```
`return nil`? If this is intended, add a comment.

tschottdorf commented on kv/txn_coord_sender.go:
```diff
@@ -865,9 +867,9 @@ func (tc *TxnCoordSender) updateState(
+				if err := tc.stopper.RunAsyncTask(func() {
 					tc.heartbeatLoop(ctx, txnID)
-				}) {
+				}); err != nil {
```
should use `err` in line 878 below?

tschottdorf commented on storage/client_test.go:
```diff
@@ -345,42 +345,47 @@ func (t *multiTestContextKVTransport) SendNext(done chan kv.BatchCall) {
+		})
+	}
+	if s == nil || runErr != nil {
 		done <- kv.BatchCall{Err: roachpb.NewSendError("store is stopped")}
```
odd to save `runErr` when you don't actually use it (maybe save `. != nil` instead, or find a way to use it)

tschottdorf commented on storage/intent_resolver.go:
```diff
@@ -417,11 +423,16 @@ func (ir *intentResolver) resolveIntents(ctx context.Context, r *Replica,
+			})
+		}
+
+		if wait || runErr != nil {
```
While you're here, add a comment to the block below that it's probably ripe for removal (including running the task).

tschottdorf commented on storage/intent_resolver.go:
```diff
@@ -440,11 +451,15 @@ func (ir *intentResolver) resolveIntents(ctx context.Context, r *Replica,
+				}
+			})
+		}
+		if wait || runErr != nil {
```
ditto.

tschottdorf commented on storage/raft_transport.go:
```diff
@@ -128,7 +128,9 @@ func (t *RaftTransport) RaftMessage(stream MultiRaft_RaftMessageServer) (err err
 		})
-	})
+	}); runErr != nil {
+		return stream.SendAndClose(new(RaftMessageResponse))
```
?

tschottdorf commented on storage/raft_transport.go:
```diff
@@ -181,11 +183,13 @@ func (t *RaftTransport) processQueue(ch chan *RaftMessageRequest, nodeID roachpb
 		})
-	})
+	}); err != nil {
+		return nil
```
`return err` or comment

tschottdorf commented on storage/replica_command.go:
```diff
@@ -1857,8 +1857,8 @@ func (r *Replica) VerifyChecksum(
-			}) {
-				log.Error("couldn't rerun consistency check as RunAsyncTask")
+			}); err != nil {
+				log.Error(err)
```
should this return?

tschottdorf commented on storage/scanner.go:
```diff
@@ -214,20 +215,24 @@ func (rs *replicaScanner) scanLoop(clock *hlc.Clock, stopper *stop.Stopper) {
-				rs.completedScan.L.Unlock()
-				if log.V(6) {
-					log.Infof("reset replica scan iteration")
+			if !shouldStop {
```
`shouldStop = nil != stopper.RunTask(....`

tschottdorf commented on storage/store.go:
```diff
@@ -921,7 +921,7 @@ func (s *Store) Start(stopper *stop.Stopper) error {
 			}
 			wg.Add(1)
-			if !s.stopper.RunLimitedAsyncTask(sem, func() {
+			if err := s.stopper.RunLimitedAsyncTask(sem, func() {
```
too much shadowing going on here. I think many of these would fare better as `if nil != ....` because the extra variable does more harm then good when it's not properly handled (tingling all kinds of spidey senses).

tschottdorf commented on storage/store.go:
```diff
@@ -953,7 +953,9 @@ func (s *Store) Start(stopper *stop.Stopper) error {
 		}
-	})
+	}); err != nil {
+		close(doneUnfreezing)
```
Good catch.

tschottdorf commented on util/stop/stopper.go:
```diff
@@ -153,71 +154,70 @@ func (s *Stopper) AddCloser(c Closer) {
 	key := taskKey{file, line}
 	if !s.runPrelude(key) {
-		return false
+		return errors.New("node stopped")
```
s/stopper/is stopping/

tschottdorf commented on util/stop/stopper.go:
```diff
@@ -153,71 +154,70 @@ func (s *Stopper) AddCloser(c Closer) {
+// as a semaphore to limit the number of tasks that are run concurrently to
+// the channel's capacity. Blocks until the semaphore is available in order to
+// push back on callers that may be trying to create many tasks. Returns an
+// error if the Stopper is draining and the function is not executed.
```
the last part of that sentence is off. `s/and the function/, in which case the function/`

tschottdorf commented on util/stop/stopper.go:
```diff
@@ -153,71 +154,70 @@ func (s *Stopper) AddCloser(c Closer) {
 	case sem <- struct{}{}:
 	case <-s.ShouldDrain():
-		return false
+		return errors.New("node stopped")
```
Make this a package-local static error.

tschottdorf commented:
Looks good, though I think that in the situations in which you're not going to do anything "standard" with the returned error, it would be better to not assign it to a variable.

For example, `if err := stopper.Stuff(); err != nil { return nil /* or something that doesn't involve err */ }` sets off alarm bells when reading. Instead, I would prefer the somewhat clumsy-looking (but clearer) `if nil != stopper.Stuff() { return nil }` (or `stopper.Stuff() != nil`, which is probably preferred since it's less backwards).

vivekmenezes commented:
Thanks for pointing that out. I've made all the necessary changes to reflect your suggestion.

---

Review status: 0 of 24 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[gossip/server.go, line 105 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sc-r1-105:-KL1LRwneh_weQsVM_3D:-639592116) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/gossip/server.go#L105)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `return nil`? If this is intended, add a comment.
</details>
Done. It's the same behavior as the 

```
case <-s.stopper.ShouldDrain():
			return nil
```

below

---

*[kv/txn_coord_sender.go, line 872 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Se-r1-872:-KL1M24EENHJCogjA1A7:-1218789955) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/kv/txn_coord_sender.go#L872)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> should use `err` in line 878 below?
</details>
Yes, makes sense

---

*[storage/client_test.go, line 389 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Si-r1-389:-KL1Mfmf0pc_n-7MqPHm:-467915377) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/client_test.go#L389)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> odd to save `runErr` when you don't actually use it (maybe save `. != nil` instead, or find a way to use it)
</details>
fixed

---

*[storage/intent_resolver.go, line 435 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sk-r1-435:-KL1NsHV2S7JzrQqs0Cb:1596185304) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/intent_resolver.go#L435)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> While you're here, add a comment to the block below that it's probably ripe for removal (including running the task).
</details>
Added a comment that it's ripe for removal.

---

*[storage/intent_resolver.go, line 462 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sk-r1-462:-KL1Nw82p_2cPse2z_dX:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/intent_resolver.go#L462)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> ditto.
</details>
done

---

*[storage/raft_transport.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sm-r1-132:-KL1OUkWWeq-wSiwonsn:1881617568) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/raft_transport.go#L132)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> ?
</details>
I now assign to _ and catch the node stopped error below

---

*[storage/raft_transport.go, line 191 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sm-r1-191:-KL1On6RB7bObLr6Ys6n:-1276660037) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/raft_transport.go#L191)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `return err` or comment
</details>
Added a comment

---

*[storage/replica_command.go, line 1861 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08So-r1-1861:-KL1PA1NqCd94kAgrRIl:-416417519) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/replica_command.go#L1861)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> should this return?
</details>
No. We want it to log the consistency check failure message below

---

*[storage/scanner.go, line 218 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Ss-r1-218:-KL1PgwTH5jHoOiTgGCe:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/scanner.go#L218)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `shouldStop = nil != stopper.RunTask(....`
</details>
done

---

*[storage/store.go, line 924 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08St-r1-924:-KL1Q0OISaWnrAd94vsI:514165753) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/store.go#L924)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> too much shadowing going on here. I think many of these would fare better as `if nil != ....` because the extra variable does more harm then good when it's not properly handled (tingling all kinds of spidey senses).
</details>
Yeah, thanks for the input

---

*[util/stop/stopper.go, line 163 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sv-r1-163:-KL1Rk3z_kHFaHNxtyVa:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/util/stop/stopper.go#L163)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/stopped/is stopping/
</details>
done

---

*[util/stop/stopper.go, line 191 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sv-r1-191:-KL1SZHqBskcZvKWMgJa:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/util/stop/stopper.go#L191)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> the last part of that sentence is off. `s/and the function/, in which case the function/`
</details>
done

---

*[util/stop/stopper.go, line 200 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sv-r1-200:-KL1RR4qVgRlU0GBBfSm:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/util/stop/stopper.go#L200)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Make this a package-local static error.
</details>
done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-:-KL1SaB_BghA4P1ny6is:1471381477)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 13 of 24 files at r1, 15 of 15 files at r2.
Review status: all files reviewed at latest revision, 24 unresolved discussions, some commit checks pending.

---

*[gossip/gossip.go, line 757 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-_IetkUSY8HcqosHT:-KL-_IetkUSY8HcqosHU:1622371738) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/gossip/gossip.go#L757)):*
> ```Go
> 				}
> 			}); err != nil {
> 				return
> ```

should this log it? gossip/inforstore.go logs.

---

*[gossip/server.go, line 105 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sc-r1-105:-KL1XQ67SW9KkYu5ey-T:-987833658) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/gossip/server.go#L105)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Done. It's the same behavior as the 
> 
> ```
> case <-s.stopper.ShouldDrain():
> 			return nil
> ```
> 
> below
</details>
I think you should return the error - that way the remote will get a descriptive error rather than just a closed connection.

---

*[kv/db.go, line 106 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1Y83Y3vnP_EWk2i0J:-KL1Y83Y3vnP_EWk2i0K:-1693726157) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/kv/db.go#L106)):*
> ```Go
> 
> 	f := func() {
> 		if err = verifyRequest(args); err != nil {
> ```

NYC but i think you can pull this out of the closure and eagerly return this `err`, which will allow you to write

```go
if err := s.stopper.RunTask(f); err != nil {
  return nil, err
}
return br, nil
```
below

---

*[storage/client_test.go, line 389 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Si-r1-389:-KL1YYOeZgRtijtI_sHd:-1871135432) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/client_test.go#L389)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> fixed
</details>
why not use it? this SendError could use it.

---

*[storage/id_alloc.go, line 98 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1YiQw725T_Fo3ljuE:-KL1YiQxBx5S2G7rXvWy:366033557) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/storage/id_alloc.go#L98)):*
> ```Go
> 						res, err = ia.db.Inc(idKey, int64(ia.blockSize))
> 					}) != nil {
> 						return
> ```

should this log?

---

*[storage/intent_resolver.go, line 435 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1Yuak0EVev1aVk09g:-KL1Yuak0EVev1aVk09h:1224795799) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/storage/intent_resolver.go#L435)):*
> ```Go
> 			// need to be resolved because they might block other tasks. See #1684.
> 			// Note that handleSkippedIntents has a TODO in case #1684 comes back.
> 			// This is ripe for removal.
> ```

can you put TODO(tschottdorf) on these?

---

*[storage/queue.go, line 394 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1Z0ZR9VyJTQRFHpOP:-KL1Z0ZSS6eYCg0lQQpj:677826560) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/storage/queue.go#L394)):*
> ```Go
> 						}
> 					}) != nil {
> 						return
> ```

log?

---

*[storage/queue.go, line 505 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1Z1j7mvl5ei5pWs95:-KL1Z1j7mvl5ei5pWs96:677826560) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/storage/queue.go#L505)):*
> ```Go
> 						}
> 					}) != nil {
> 						return
> ```

log?

---

*[storage/raft_transport.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sm-r1-132:-KL1Z6M7jjOTpr1UwqJI:1816242356) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/raft_transport.go#L132)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I now assign to _ and catch the node stopped error below
</details>
i think you should return this error

---

*[storage/raft_transport.go, line 191 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sm-r1-191:-KL1Z95rhsqRYZJa09X8:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/raft_transport.go#L191)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Added a comment
</details>
ditto

---

*[storage/replica.go, line 2300 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1ZF1Rvr1ROH-HBrdP:-KL1ZF1Rvr1ROH-HBrdQ:888570080) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/storage/replica.go#L2300)):*
> ```Go
> 	var hasLease bool
> 	var pErr *roachpb.Error
> 	if runErr := r.store.Stopper().RunTask(func() {
> ```

i think you can call this `err`

---

*[storage/replica_command.go, line 1861 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08So-r1-1861:-KL1ZQ4GMrBUX5-vISQ7:706573452) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/replica_command.go#L1861)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> No. We want it to log the consistency check failure message below
</details>
i think this wants some context: `("couldn't rerun consistency check as RunAsyncTask: %s", err)`

---

*[util/stop/stopper.go, line 31 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1X5et7pmsfLQ_3oKB:-KL1X5et7pmsfLQ_3oKC:1292615664) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/util/stop/stopper.go#L31)):*
> ```Go
> )
> 
> var errNodeStopping = errors.New("node is stopping")
> ```

this doesn't really have anything to do with a node, so the text should be more agnostic.

---

*[util/stop/stopper_test.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-ZY7jqh1MM7aY4QPC:-KL-ZY7jqh1MM7aY4QPD:1696215072) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/util/stop/stopper_test.go#L194)):*
> ```Go
> 			// Wait until Quiesce() is called.
> 			<-qc
> 			if err := thisStopper.RunTask(func() {}); err == nil {
> ```

should check for the specific error

---

*[util/stop/stopper_test.go, line 414 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-ZgVQQr4jg5MNgbxm:-KL-ZgVQQr4jg5MNgbxn:-95600408) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/util/stop/stopper_test.go#L414)):*
> ```Go
> 	for i := 0; i < maxConcurrency*3; i++ {
> 		wg.Add(1)
> 		_ = s.RunLimitedAsyncTask(sem, f)
> ```

why not check the error?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 15 of 24 files reviewed at latest revision, 24 unresolved discussions, some commit checks pending.

---

*[gossip/gossip.go, line 757 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-_IetkUSY8HcqosHT:-KL24Enq7tPaufBMcz1J:1730422223) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/gossip/gossip.go#L757)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> should this log it? gossip/inforstore.go logs.
</details>
I don't think it's worth it. Just as much as we don't log down below when we exit the select.

---

*[gossip/server.go, line 105 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sc-r1-105:-KL25JT2-eI_gLco7Bxz:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/gossip/server.go#L105)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think you should return the error - that way the remote will get a descriptive error rather than just a closed connection.
</details>
done

---

*[kv/db.go, line 106 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1Y83Y3vnP_EWk2i0J:-KL28lZfxQc99uaBy7Jq:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/kv/db.go#L106)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> NYC but i think you can pull this out of the closure and eagerly return this `err`, which will allow you to write
> 
> ```go
> if err := s.stopper.RunTask(f); err != nil {
>   return nil, err
> }
> return br, nil
> ```
> below
</details>
done

---

*[storage/client_test.go, line 389 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Si-r1-389:-KL29OTuNzt0sD0JNE0j:-1436512859) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/client_test.go#L389)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why not use it? this SendError could use it.
</details>
I'm leaving this as is because it also covers the s == nil case

---

*[storage/id_alloc.go, line 98 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1YiQw725T_Fo3ljuE:-KL29gMm2dvfuqiCw4k9:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/storage/id_alloc.go#L98)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> should this log?
</details>
done

---

*[storage/intent_resolver.go, line 435 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1Yuak0EVev1aVk09g:-KL2A9ZxBl0_3TvX72nv:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/storage/intent_resolver.go#L435)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you put TODO(tschottdorf) on these?
</details>
done

---

*[storage/queue.go, line 394 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1Z0ZR9VyJTQRFHpOP:-KL2AHCiF4Caw0w5wAvg:-398301746) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/storage/queue.go#L394)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> log?
</details>
the ShouldStop() above doesn't so I'm leaving as is

---

*[storage/queue.go, line 505 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1Z1j7mvl5ei5pWs95:-KL2ASp_5U_eUlrRkv2V:-1110202927) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/storage/queue.go#L505)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> log?
</details>
same ShouldStop() doesn't log

---

*[storage/raft_transport.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sm-r1-132:-KL2B-t2ATuKe2NMj0hE:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/raft_transport.go#L132)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i think you should return this error
</details>
done

---

*[storage/raft_transport.go, line 191 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08Sm-r1-191:-KL2BFt4q44L3NY0Vh1m:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/raft_transport.go#L191)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
done

---

*[storage/replica.go, line 2300 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1ZF1Rvr1ROH-HBrdP:-KL2B_kkSHPKQ64lWsOQ:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/storage/replica.go#L2300)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i think you can call this `err`
</details>
done

---

*[storage/replica_command.go, line 1861 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08So-r1-1861:-KL2Bx1ZtZseT2v7ZvJb:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/replica_command.go#L1861)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i think this wants some context: `("couldn't rerun consistency check as RunAsyncTask: %s", err)`
</details>
done

---

*[util/stop/stopper.go, line 31 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL1X5et7pmsfLQ_3oKB:-KL2CT7ha6RvT1wYcoGQ:-326205601) ([raw file](https://github.com/cockroachdb/cockroach/blob/f71e7b9cd30b62522d205ee8e50b3c7e85b2b748/util/stop/stopper.go#L31)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this doesn't really have anything to do with a node, so the text should be more agnostic.
</details>
changed to, "stopper is stopping"

---

*[util/stop/stopper_test.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-ZY7jqh1MM7aY4QPC:-KL2DYqJ0SHq36rjNVtX:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/util/stop/stopper_test.go#L194)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> should check for the specific error
</details>
done

---

*[util/stop/stopper_test.go, line 414 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-ZgVQQr4jg5MNgbxm:-KL2DkKB3BztLmx_cl6I:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/util/stop/stopper_test.go#L414)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why not check the error?
</details>
done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 9 of 9 files at r3.
Review status: all files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[storage/replica_command.go, line 1861 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08So-r1-1861:-KL2I8O7nLBZZyKv4KEE:-1304484994) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/replica_command.go#L1861)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> done
</details>
s/ as/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 20 of 24 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[storage/replica_command.go, line 1861 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL-W-LV4acOzeXm08So-r1-1861:-KL2Uci09ed1RutBgQWd:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/storage/replica_command.go#L1861)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/ as/
</details>
done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 24 of 24 files at r1, 15 of 15 files at r2, 9 of 9 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[kv/db.go, line 106 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL2ZK75q76xxWRrmKbx:-KL2ZK75q76xxWRrmKby:-209640971) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d25b6e93006a08b5eebc69d0591086c084958b/kv/db.go#L106)):*
> ```Go
> 
> 	if err = verifyRequest(args); err != nil {
> 		return
> ```

No need to use all of naked returns. Instead

```go
if err := .... {
   return nil, err
}
```

---

*[kv/db.go, line 109 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL2ZQ48wptOTlFFK0GD:-KL2ZQ48wptOTlFFK0GE:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d25b6e93006a08b5eebc69d0591086c084958b/kv/db.go#L109)):*
> ```Go
> 	}
> 
> 	err = s.stopper.RunTask(func() {
> ```

ditto

---

*[kv/txn_coord_sender.go, line 878 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL2Y_DQEpELYTJ0Dd5x:-KL2Y_DQEpELYTJ0Dd5y:-1456760710) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/kv/txn_coord_sender.go#L878)):*
> ```Go
> 					// In principle, we can relax this as needed though.
> 					tc.unregisterTxnLocked(txnID)
> 					return roachpb.NewError(&roachpb.NodeUnavailableError{})
> ```

Was NodeUnavailableError treated specially anywhere? If so, this is worth a second look. If not, maybe NodeUnavailableError can be removed from the codebase.

---

*[storage/replica_command.go, line 1861 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL2ZhqrbjKgiVRFSt9H:-KL2ZhqrbjKgiVRFSt9I:232312881) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d25b6e93006a08b5eebc69d0591086c084958b/storage/replica_command.go#L1861)):*
> ```Go
> 				}
> 			}); err != nil {
> 				log.Errorf("couldn't rerun consistency check as: %s", err)
> ```

Somewhat unrelated, but shouldn't we use `log.Error(errors.Cause("could not rerun consistency check"), err)`?

---

*[util/stop/stopper.go, line 31 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL2Zyh2e661jrqox5pV:-KL2Zyh2e661jrqox5pW:1668786046) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d25b6e93006a08b5eebc69d0591086c084958b/util/stop/stopper.go#L31)):*
> ```Go
> )
> 
> var errStopperStopping = errors.New("stopper is stopping")
> ```

Don't undo if someone asked you to do it, but `errStopping` sounds about as good and doesn't stutter.

---

*[util/stop/stopper_test.go, line 196 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL2_4qLg5yij8zlIcWY:-KL2_4qMHT1cUurWHAlA:-381767038) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d25b6e93006a08b5eebc69d0591086c084958b/util/stop/stopper_test.go#L196)):*
> ```Go
> 			<-qc
> 			if err := thisStopper.RunTask(func() {}); !testutils.IsError(
> 				err, "stopper is stopping",
> ```

use `errStopperStopping.String()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
All leftover comments from this PR are resolved in https://github.com/cockroachdb/cockroach/pull/7479

---

Review status: all files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[kv/db.go, line 106 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL2ZK75q76xxWRrmKbx:-KLEJhShxzCgIE-Xjlh1:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d25b6e93006a08b5eebc69d0591086c084958b/kv/db.go#L106)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> No need to use all of naked returns. Instead
> 
> ```go
> if err := .... {
>    return nil, err
> }
> ```
</details>
done

---

*[kv/db.go, line 109 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL2ZQ48wptOTlFFK0GD:-KLEJibO3NCDn14cpy5R:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d25b6e93006a08b5eebc69d0591086c084958b/kv/db.go#L109)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> ditto
</details>
done

---

*[kv/txn_coord_sender.go, line 878 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL2Y_DQEpELYTJ0Dd5x:-KLEJs9ZUfd7329_S1Y0:-629249585) ([raw file](https://github.com/cockroachdb/cockroach/blob/222bff830bc336fd3e7bac87d70dfb2d2be6cf7c/kv/txn_coord_sender.go#L878)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Was NodeUnavailableError treated specially anywhere? If so, this is worth a second look. If not, maybe NodeUnavailableError can be removed from the codebase.
</details>
Its used elsewhere in another location and used for retries through

```
// handlePerReplicaError returns true if the given error is likely to
// be unique to the replica that reported it, and retrying on other
// replicas is likely to produce different results.
```

---

*[storage/replica_command.go, line 1861 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL2ZhqrbjKgiVRFSt9H:-KLEKf421W0qD0CxyxvS:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d25b6e93006a08b5eebc69d0591086c084958b/storage/replica_command.go#L1861)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Somewhat unrelated, but shouldn't we use `log.Error(errors.Cause("could not rerun consistency check"), err)`?
</details>
done

---

*[util/stop/stopper.go, line 31 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL2Zyh2e661jrqox5pV:-KLEKoHsWiF2IpCNyPKy:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d25b6e93006a08b5eebc69d0591086c084958b/util/stop/stopper.go#L31)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Don't undo if someone asked you to do it, but `errStopping` sounds about as good and doesn't stutter.
</details>
done

---

*[util/stop/stopper_test.go, line 196 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-KL2_4qLg5yij8zlIcWY:-KLEKp0VBmwqA0PFKFDZ:1769476350) ([raw file](https://github.com/cockroachdb/cockroach/blob/84d25b6e93006a08b5eebc69d0591086c084958b/util/stop/stopper_test.go#L196)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> use `errStopperStopping.String()`.
</details>
Its in a different package and I didn't feel like exporting it for this one use.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7459#-:-KLELZSUZHhNOIlRlAsU:-1823698545)*
<!-- Sent from Reviewable.io -->
